src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs

```diff
@@ -245,72 +245,52 @@ void addArg(RefKind refKind, BoundExpression expression)
-                            var outputTemp = new BoundDagTemp(e.Syntax, e.IndexerType, e);
-                            BoundExpression output = _tempAllocator.GetTemp(outputTemp);
-                            return _factory.AssignmentExpression(output, access);
-                        }
-
-                    case BoundDagIndexerEvaluation { IndexerSymbol: PropertySymbol indexer } e:
+                    case BoundDagIndexerEvaluation e:
                         {
+                            // this[Index]
                             // this[int]
-                            BoundExpression access = _factory.Indexer(input, indexer, makeImplicitIndexArgument(e.Index, e.LengthTemp));
-                            var outputTemp = new BoundDagTemp(e.Syntax, e.IndexerType, e);
-                            BoundExpression output = _tempAllocator.GetTemp(outputTemp);
-                            return _factory.AssignmentExpression(output, access);
-                        }
+                            // array[Index]
 
-                    case BoundDagIndexerEvaluation { IndexerAccess: BoundIndexerAccess indexerAccess } e:
-                        {
-                            // this[Index]
-                            BoundExpression access = makeIndexerAccess(e.Syntax, indexerAccess, makeExplicitIndexArgument(e.Index, e.Index < 0));
-                            var outputTemp = new BoundDagTemp(e.Syntax, e.IndexerType, e);
-                            BoundExpression output = _tempAllocator.GetTemp(outputTemp);
-                            return _factory.AssignmentExpression(output, access);
-                        }
+                            var indexerAccess = e.IndexerAccess;
+                            if (indexerAccess is BoundIndexOrRangePatternIndexerAccess implicitAccess)
+                            {
+                                indexerAccess = implicitAccess.WithLengthOrCountAccess(_tempAllocator.GetTemp(e.LengthTemp));
+                            }
 
-                    case BoundDagSliceEvaluation { IndexerAccess: null, SliceMethod: null } e:
-                        {
-                            // RuntimeHelpers.GetSubArray(T[], Range)
-                            TypeSymbol inputType = input.Type;
-                            Debug.Assert(inputType is { });
-                            Debug.Assert(inputType.IsSZArray());
-                            Debug.Assert(e.StartIndex >= 0 && e.EndIndex <= 0);
+                            var placeholderValues = PooledDictionary<BoundValuePlaceholderBase, BoundExpression>.GetInstance();
+                            placeholderValues.Add(e.ReceiverPlaceholder, input);
+                            placeholderValues.Add(e.ArgumentPlaceholder, makeUnloweredIndexArgument(e.Index));
+                            indexerAccess = PlaceholderReplacer.Replace(placeholderValues, indexerAccess);
```

Comment:
:memo: I ended up opting for a rewriter because the number of shapes that we needed to handle was growing (need to handle implicit conversions to System.Index/Range for example).

Reply:
curious why we need to do this, aren't placeholders ample to do the replacement?

Reply:
It is not absolutely required, but if we were to use the LocalRewriter to replace those placeholders, they would behave differently than other placeholders. The placeholders would be mapped to unlowered nodes, then LocalRewriter would have to visit the nodes during substitution.